core/domain/stats_jobs_one_off.py

```diff
@@ -127,3 +143,2369 @@ def reduce(key, stringified_values):
+        if rule_str != cls._DEFAULT_RULESPEC_STR:
+            if rule_spec.rule_type != 'OutputEquals' and (
+                    rule_spec.rule_type not in rule_types_without_output):
+                return (
+                    None,
+                    'Cannot reconstitute a CodeEvaluation object without '
+                    'OutputEquals, CodeContains, CodeDoesNotContain, or '
+                    'ResultsInError rules.')
+            if rule_spec.rule_type == 'OutputEquals':
+                code_output = rule_spec.inputs['x']
+                code_evaluation_dict = {
+                    'code': answer_str,
+                    'output': code_output,
+                    'evaluation': '',
+                    'error': ''
+                }
+                return cls._normalize_raw_answer_object(
+                    objects.CodeEvaluation, code_evaluation_dict, answer_str)
+
+        # Otherwise the answer is a default answer, or the rule type is one of
+        # CodeContains, CodeDoesNotContain, or ResultsInError.
+        code_evaluation_dict = {
+            'code': answer_str,
+            'output': '',
+            'evaluation': '',
+            'error': ''
+        }
+        return cls._normalize_raw_answer_object(
+            objects.CodeEvaluation, code_evaluation_dict, answer_str)
+
+    @classmethod
+    def _cb_reconstitute_continue(
+            cls, state, answer_group_index, rule_spec, rule_str, answer_str):
+        # The Jinja representation for CodeEvaluation answer strings is blank.
+        if not rule_spec and not answer_str and (
+                rule_str == cls._DEFAULT_RULESPEC_STR):
+            # There is no answer for 'Continue' interactions.
+            return (None, None)
+        return (
+            None,
+            'Expected Continue submissions to only be default rules: %s'
+            % rule_str)
+
+    @classmethod
+    def _cb_reconstitute_end_exploration(
+            cls, state, answer_group_index, rule_spec, rule_str, answer_str):
+        return (
+            None,
+            'There should be no answers submitted for EndExploration.')
+
+    @classmethod
+    def _cb_reconstitute_graph_input(
+            cls, state, answer_group_index, rule_spec, rule_str, answer_str):
+        # pylint: disable=line-too-long
+        # The Jinja representation for Graph answer strings is:
+        #   ({% for vertex in answer.vertices -%}
+        #     {% if answer.isLabeled -%}{{vertex.label}}{% else -%}{{loop.index}}{% endif -%}
+        #     {% if not loop.last -%},{% endif -%}
+        #   {% endfor -%})
+        #   [{% for edge in answer.edges -%}
+        #     ({{edge.src}},{{edge.dst}}){% if not loop.last -%},{% endif -%}
+        #   {% endfor -%}]
+
+        # This answer type is not being reconstituted. 'HasGraphProperty' has
+        # never had an answer submitted for it. 'IsIsomorphicTo' has had 5
+        # answers submitted for it, 4 of which are too long to actually
```

Comment:
It contributes to the overall length of an answer which can't exceed 1 entity of the sharded data store, but only for the migration. After the migration is done, we don't need to store the rule spec string in the answer structure (and we don't). If we ever need the rule spec string corresponding to an answer, we can run a MR job and lookup the exact state (and version) that was used to generate the answer in the first place, along with the answer group index and rule spec index.